kyo-data/shared/src/main/scala/kyo/Ansi.scala

```diff
@@ -1,58 +1,112 @@
+
+        /** Applies grey color to the string. */
+        def grey: String = s"\u001b[90m$str\u001b[0m"
+
+        /** Applies bold formatting to the string. */
+        def bold: String = s"\u001b[1m$str\u001b[0m"
+
+        /** Applies dim formatting to the string. */
+        def dim: String = s"\u001b[2m$str\u001b[0m"
+
+        /** Applies italic formatting to the string. */
+        def italic: String = s"\u001b[3m$str\u001b[0m"
+
+        /** Applies underline formatting to the string. */
         def underline: String = s"\u001b[4m$str\u001b[0m"
 
+        /** Removes all ANSI escape sequences from the string. */
         def stripAnsi: String = str.replaceAll("\u001b\\[[0-9;]*[a-zA-Z]", "")
     end extension
 
     object highlight:
+
+        /** Applies syntax highlighting to a code snippet with optional header and trailer.
+          *
+          * @param header
+          *   The header text to be displayed before the code (optional).
+          * @param code
+          *   The main code snippet to be highlighted.
+          * @param trailer
+          *   The trailer text to be displayed after the code (optional).
+          * @param startLine
+          *   The starting line number for the code snippet (default is 1).
+          * @return
+          *   A string with the highlighted code, including line numbers and formatting.
+          */
         def apply(header: String, code: String, trailer: String, startLine: Int = 1): String =
-            val headerLines  = if header.nonEmpty then header.split("\n") else Array.empty[String]
-            val codeLines    = code.split("\n").dropWhile(_.trim.isEmpty).reverse.dropWhile(_.trim.isEmpty).reverse
-            val trailerLines = if trailer.nonEmpty then trailer.split("\n") else Array.empty[String]
-
-            val allLines        = headerLines ++ codeLines ++ trailerLines
-            val toDrop          = codeLines.filter(_.trim.nonEmpty).map(_.takeWhile(_ == ' ').length).minOption.getOrElse(0)
-            val lineNumberWidth = (startLine + codeLines.length - 1).toString.length
-            val separator       = "│".dim
-
-            val processedLines = allLines.zipWithIndex.map { case (line, index) =>
-                val isHeader  = index < headerLines.length
-                val isTrailer = index >= (headerLines.length + codeLines.length)
-                val lineNumber =
-                    if isHeader || isTrailer then
-                        " ".repeat(lineNumberWidth)
-                    else
-                        (startLine + index - headerLines.length).toString.padTo(lineNumberWidth, ' ')
-
-                val highlightedLine =
-                    if isHeader || isTrailer then
-                        line.green
-                    else
-                        highlightLine(line.drop(toDrop))
-
-                s"${lineNumber.dim} $separator $highlightedLine"
-            }
-
-            processedLines.mkString("\n")
+            try
```

Comment:
An unrelated improvement: this method is now used in several places to report errors. If it throws an exception, we'd lose visibility of the underlying failure so I've added a fallback to return the code without highlighting if that happens.